lib/url-decorator.jsx

```diff
@@ -0,0 +1,53 @@
+import React from 'react';
+import { CompositeDecorator } from 'draft-js';
+import urlRegex from 'url-regex';
+import normalizeUrl from 'normalize-url';
+import { viewExternalUrl } from './utils/url-utils';
+
+const urlRe = urlRegex();
+/*
+- markdownLinkRe is a regexp to be applied to the substring preceding a matched url, ie. `wholeString.slice(0, wholeString.indexOf(matchedUrl))`
+- it tests whether the matched url is preceded by the charcters that define a link in markdown, ie. `(` or `]: `
+*/
+const markdownLinkRe = /(?:\(|\]:\s+)$/;
+
+export function normalizeUrlsInMarkdownString( str ) {
+	return str.replace( urlRe, ( match, offset ) => {
+		var normalized = normalizeUrl( match );
+		if ( markdownLinkRe.test( str.slice( 0, offset ) ) ) {
+			return normalized;
+		}
+		// replace url with markdown link only if it's not already part of one
+		return `[${match}](${normalized})`;
+	} );
+}
```

Comment:
I agree, this version of this function definitely wasn't one of my best creations.

how about this?

```js
const mdRe = new RegExp(
	'(\\(|\\]:\\s+)?(' + urlRe.source + ')',
	urlRe.flags
);

export function normalizeUrlsInMarkdownString( str ) {
	return str.replace( mdRe, function( _, markdownLinkChars, url ) {
		var normalizedUrl = normalizeUrl( url );

		if ( markdownLinkChars !== undefined ) {
			// if the first capturing group of the regex matched
			// then this url is part of a markdown link
			return markdownLinkChars + normalizedUrl;
		}

		// replace the matched url with a markdown link, to make it clickable
		return `[${url}](${normalizedUrl})`;
	} )
}
```

basically this is modifying `urlRe` to also capture either `(` or `]: ` if it appears right before the url. If either one is there it means the url is part of a markdown link so we shouldn't replace it with a markdown link to make it clickable because it is already clickable. makes sense? (this is definitely much faster)